File: tri.py
"""FaceGen TRI (FRTRI003) morph file parser — Skyrim expression/phoneme morphs."""
import struct


class TriFile:
    def __init__(self, data):
        assert data[:8] == b'FRTRI003', data[:8]
        (self.num_verts, num_tris, num_quads, unk1, unk2, num_uv, flags,
         num_morphs, num_mods, num_mod_verts) = struct.unpack_from('<10i', data, 8)
        o = 64
        self.verts = struct.unpack_from('<%df' % (self.num_verts * 3), data, o)
        o += self.num_verts * 12
        o += num_mod_verts * 12
        o += num_tris * 12  # triangle indices
        if flags & 1:
            o += num_uv * 8       # uv coords
            o += num_tris * 12    # uv indices
        self.morphs = {}  # name -> list of (dx,dy,dz) per vertex
        for _ in range(num_morphs):
            nlen = struct.unpack_from('<I', data, o)[0]
            o += 4
            name = data[o:o + nlen].rstrip(b'\x00').decode('latin-1')
            o += nlen
            scale = struct.unpack_from('<f', data, o)[0]
            o += 4
            raw = struct.unpack_from('<%dh' % (self.num_verts * 3), data, o)
            o += self.num_verts * 6
            self.morphs[name] = [(raw[i] * scale, raw[i+1] * scale, raw[i+2] * scale)
                                 for i in range(0, len(raw), 3)]
        self.trailing = len(data) - o


if __name__ == '__main__':
    import sys
    from bsa import BSA
    if sys.argv[1].endswith('.bsa'):
        data = BSA(sys.argv[1]).read(sys.argv[2])
    else:
        data = open(sys.argv[1], 'rb').read()
    t = TriFile(data)
    print(f'verts={t.num_verts} morphs={len(t.morphs)} trailing_bytes={t.trailing}')
    for name, deltas in t.morphs.items():
        mx = max(abs(c) for v in deltas for c in v)
        print(f'  {name}: max_delta={mx:.3f}')

File: internal/tri/tri.go
// Package tri parses game FaceGen FRTRI003 morph files.
package tri

import (
	"fmt"
	"strings"

	"soulgem/internal/binread"
	"soulgem/internal/mathutil"
)

type File struct {
	NumVerts int
	Verts    []float32
	Morphs   map[string][]mathutil.Vec3
	Order    []string
	Trailing int
}

// header is the fixed 64-byte FRTRI003 header. Only the counts the morph reader
// needs are named; the rest is padding the format reserves.
type header struct {
	numVerts    int
	numTris     int
	numUV       int
	flags       int32
	numMorphs   int
	numModVerts int
}

func readHeader(c *binread.Cursor) header {
	c.Seek(8) // past the "FRTRI003" signature
	h := header{numVerts: int(c.I32()), numTris: int(c.I32())}
	c.Skip(12) // quads and two reserved counts
	h.numUV = int(c.I32())
	h.flags = c.I32()
	h.numMorphs = int(c.I32())
	c.Skip(4) // mods
	h.numModVerts = int(c.I32())
	c.Seek(64)
	return h
}

func Parse(data []byte) (*File, error) {
	if len(data) < 64 || string(data[:8]) != "FRTRI003" {
		return nil, fmt.Errorf("unsupported TRI signature %q", data[:min(8, len(data))])
	}
	c := binread.New("TRI", data)
	h := readHeader(c)

	out := &File{NumVerts: h.numVerts, Morphs: make(map[string][]mathutil.Vec3)}
	out.Verts = make([]float32, 3*c.Count(h.numVerts, 12, "vertex"))
	for i := range out.Verts {
		out.Verts[i] = float32(c.F32())
	}
	c.Skip(h.numModVerts*12 + h.numTris*12)
	if h.flags&1 != 0 {
		c.Skip(h.numUV*8 + h.numTris*12)
	}
	// Each morph is a name, a scale, and one int16 triple per vertex.
	for range c.Count(h.numMorphs, 4+2+6*h.numVerts, "morph") {
		name := strings.TrimRight(c.SizedString(), "\x00")
		scale := c.F32()
		deltas := make([]mathutil.Vec3, c.Count(h.numVerts, 6, "morph delta"))
		for i := range deltas {
			deltas[i] = mathutil.Vec3{
				float64(c.I16()) * scale,
				float64(c.I16()) * scale,
				float64(c.I16()) * scale,
			}
		}
		if c.Failed() {
			return nil, c.Err()
		}
		out.Morphs[name] = deltas
		out.Order = append(out.Order, name)
	}
	if err := c.Err(); err != nil {
		return nil, err
	}
	out.Trailing = len(data) - c.Off()
	return out, nil
}
